### PR TITLE
Testing that Netty clients fail for mismatching TLS common names

### PR DESCRIPTION
**Pull request by hstefan**
Netty is known to not use SSL hostname verfication by default[1], although this is enabled for gRPC channels. This change validates that functionality by porting a test with a similar purpuse for OkHttp.

As benefit to users it helps build confidence that their clients aren't vulnerable to Improper Certificate Validation issues reported by vulnerability scanners such as Snyk[2].

* [1] https://github.com/netty/netty/issues/8537
* [2] https://security.snyk.io/vuln/SNYK-JAVA-IONETTY-1042268

### CODE REVIEW

**Comment by ejona86** [comment]
The CLA needs to be agreed to for me to look at the code, but I'd imagine TlsTest is the best place for this, because at the end of the day this logic is performed by the SSL provider and that test is run for each SSL provider.

It'd basically be a copy of `basicClientServerIntegrationTest`, except `clientChannel()` would be inlined and [`overrideAuthority()`](https://github.com/grpc/grpc-java/blob/881747a63df79a7baae252de86447c0997c76f69/netty/src/test/java/io/grpc/netty/TlsTest.java#L355) would use a random (wrong) server name.

**Comment by hstefan** [comment]
@ejona86 thanks for the suggestion, I reworked my PR around that idea.

CLA should also be all set now :-)

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 1]
```diff
@@ -185,6 +187,52 @@ public void basicClientServerIntegrationTest() throws Exception {
...
+              .trustManager(clientTrustedCaCerts)
+              .build()
+            )
+            .build();
+    SimpleServiceGrpc.SimpleServiceBlockingStub client = SimpleServiceGrpc.newBlockingStub(channel);
+
+    // Send an actual request, via the full GRPC & network stack, and check that a proper
+    // response comes back.
+    try {
+      client.unaryRpc(SimpleRequest.getDefaultInstance());
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:229]: A `fail()` is needed in the try after `unaryRpc()`, otherwise if the RPC succeeds the test will still pass. (This is why the CI failed, because there's a checker for mistakes like that.)

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 2]
```diff
@@ -49,6 +49,8 @@
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLHandshakeException;
+
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:53]: nit: We don't separate these into sections. You can remove this extra line.

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 3]
```diff
@@ -185,6 +187,52 @@ public void basicClientServerIntegrationTest() throws Exception {
...
+    // Override authority on the client to trigger a hostname verification failure
+    channel = NettyChannelBuilder.forAddress("localhost", server.getPort())
+            .overrideAuthority("i.am.a.bad.hostname")
+            .negotiationType(NegotiationType.TLS)
+            .sslContext(clientContextBuilder
+              .keyManager(clientCertChainFile, clientPrivateKeyFile)
+              .trustManager(clientTrustedCaCerts)
+              .build()
+            )
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:222]: nit: our style never has whitespace before a parenthesis.

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 4]
```diff
@@ -75,8 +76,8 @@
    */
   @Parameters(name = "{0}")
   public static Iterable<Object[]> data() {
-    return Arrays.asList(new Object[][] {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:78]: It seems you've run an autoformatter on the code. Please revert the changes to existing code that you aren't already touching.

Comment (hstefan) [reply netty/src/test/java/io/grpc/netty/TlsTest.java]: Should be fixed now :-)

**Review by ejona86 — APPROVED** [review]

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 5]
```diff
@@ -185,6 +186,54 @@ public void basicClientServerIntegrationTest() throws Exception {
     client.unaryRpc(SimpleRequest.getDefaultInstance());
   }
 
+  /**
+   * Test that a client fails to send requests over TLS to a server with an unexpected hostname
+   * listed in an otherwise valid certificate.
+   */
+  @Test
+  public void serverRejectedBadHostname() throws Exception {
```
Comment (sanjaypujare) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:194]: Client does the rejection. Can we rename to `clientRejectedMismatchedHostname` ?

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 6]
```diff
@@ -185,6 +186,54 @@ public void basicClientServerIntegrationTest() throws Exception {
...
+        .build();
+    SimpleServiceGrpc.SimpleServiceBlockingStub client = SimpleServiceGrpc.newBlockingStub(channel);
+
+    // Send an actual request, via the full GRPC & network stack, and check that a proper
+    // response comes back.
+    try {
+      client.unaryRpc(SimpleRequest.getDefaultInstance());
+      fail("TLS handshake should have failed, but didn't; received RPC response");
+    } catch (StatusRuntimeException e) {
+      assertEquals(Throwables.getStackTraceAsString(e), Status.Code.UNAVAILABLE,
```
Comment (sanjaypujare) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:231]: The 1st argument seems completely unnecessary. We can change this to the Truth version as follows:
```
assertThat(e.getStatus().getCode()).isEqualTo(Status.Code.UNAVAILABLE);
```

File: netty/src/test/java/io/grpc/netty/TlsTest.java [hunk 7]
```diff
@@ -185,6 +186,54 @@ public void basicClientServerIntegrationTest() throws Exception {
...
+          e.getStatus().getCode());
+      assertThat(e.getCause()).isInstanceOf(SSLHandshakeException.class);
```
Comment (sanjaypujare) [inline netty/src/test/java/io/grpc/netty/TlsTest.java:233]: Can we also assert the root cause is certificateException as
```
assertThat(e.getCause().getCause()).isInstanceOf(CertificateException.class);
```
Also it will be nice (but not necessary) to check the message in the `CertificateException` is something like "subject alternative DNS name mismatch...".

File: netty/src/test/java/io/grpc/netty/TlsTest.java (hunk 6 above)
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/TlsTest.java:231]: The current line produces a much better error message when it fails.

File: netty/src/test/java/io/grpc/netty/TlsTest.java (hunk 7 above)
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/TlsTest.java:233]: FWIW, I suspect that isn't stable across the various security providers, so isn't feasible.

Comment (sanjaypujare) [reply netty/src/test/java/io/grpc/netty/TlsTest.java:233]: That's what I initially suspected as well but I did a test and for all 3 providers the assertion is true.

File: netty/src/test/java/io/grpc/netty/TlsTest.java (hunk 6 above)
Comment (sanjaypujare) [reply netty/src/test/java/io/grpc/netty/TlsTest.java:231]: I see that pattern used elsewhere in the test file. Makes sense.

File: netty/src/test/java/io/grpc/netty/TlsTest.java (hunk 7 above)
Comment (sanjaypujare) [reply netty/src/test/java/io/grpc/netty/TlsTest.java:233]: It is best to make sure the SSLHandshake exception is due to hostname mismatch and is not a false positive

**Comment by hstefan** [comment]
@sanjaypujare thanks for the reviewed, please have another look when I have time.

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Comment by sanjaypujare** [comment]
> @sanjaypujare thanks for the review, please have another look when you have time.

Just approved.

**Comment by ejona86** [comment]
This got lost in my open tabs. The CI shows there were some failures on other platforms, but it should be as simply as allowing other Exceptions. But we've seen some TLS stuff change recently with a Netty upgrade, so re-running to see if that's still accurate.
